[SPARK-37829][SQL] DataFrame.joinWith should return null rows for missing values

cdegroc opened the pull request:
### What changes were proposed in this pull request?
We add a unit test demonstrating a regression on `DataFrame.joinWith` and fix the regression by updating `ExpressionEncoder`. The fix is equivalent to reverting [this commit](https://github.com/apache/spark/commit/cd92f25be5a221e0d4618925f7bc9dfd3bb8cb59).

### Why are the changes needed?
Doing an outer-join using joinWith on DataFrames used to return missing values as null in Spark 2.4.8, but returns them as Rows with null values in Spark 3.0.0+.

The regression has been introduced in [this commit](https://github.com/apache/spark/commit/cd92f25be5a221e0d4618925f7bc9dfd3bb8cb59).

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added a unit test. This unit test succeeds with Spark 2.4.8 but fails with Spark 3.0.0+.

The new unit test does a left outer join on two DataFrames using the `joinWith` method.
The join is performed on the `b` field of `ClassData` (Ints).
The row `ClassData("a", 1)` on the left side of the join has no matching row on the right side of the join as there is no row with value `1` for field `b`.
The missing value (of Row type) is represented as a `GenericRowWithSchema(Array(null, null), rightFieldSchema)` instead of a `null` value making the test fail.

This new test is identical to [this one](https://github.com/apache/spark/blob/6061e60ed7691bf3ca0e0964acac4c53c69bcc07/sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala#L1269-L1292) and only differs in that it uses DataFrames instead of Datasets.

I've run unit tests for the `sql-core` and `sql-catalyst` submodules locally with `./build/mvn clean package -pl sql/core,cql/catalyst`

cdegroc committed:
[SPARK-37829][SQL][TESTS] Add a test demonstrating joinWith outer join issue

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented:
cc @cloud-fan @viirya FYI

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:
```diff
@@ -117,9 +117,14 @@ object ExpressionEncoder {
         case GetColumnByOrdinal(0, _) => input
       }
+      if (enc.objSerializer.nullable) {
+        If(IsNull(input), Literal.create(null, newDeserializer.dataType), newDeserializer)
```
Can we create a new JIRA ticket and PR for this bug?

cdegroc replied:
👋  I've already created [SPARK-37829](https://issues.apache.org/jira/browse/SPARK-37829). What do you mean exactly?

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -2147,6 +2148,50 @@ class DatasetSuite extends QueryTest
   }
+
+  test("SPARK-37829: DataFrame outer join") {
+    // Same as "SPARK-15441: Dataset outer join" but using DataFrames instead of Datasets
```
This comment looks unnecessary and confusing. We can remove it.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:
```diff
@@ -117,9 +117,14 @@ object ExpressionEncoder {
         case GetColumnByOrdinal(0, _) => input
       }
+      if (enc.objSerializer.nullable) {
+        If(IsNull(input), Literal.create(null, newDeserializer.dataType), newDeserializer)
```
Can we reuse `nullSafe`?

cloud-fan commented:
> Add a unit test demonstrating the regression on DataFrame.joinWith.

Let's be more clear about "What changes were proposed in this pull request?". I thought this is a test-only PR but it actually fixed a bug. Please describe the test in `How was this patch tested?` section.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:
```diff
@@ -110,23 +110,28 @@ object ExpressionEncoder {
 
       val input = GetStructField(newDeserializerInput, index)
-      enc.objDeserializer.transformUp {
+      val newDeserializer = enc.objDeserializer.transformUp {
```
do you mean the `objDeserializer` can be something that doesn't propagate nulls, so we need to manually check null input and create null literal? If so please add some code comments to explain it.

cdegroc replied:
It seems so. I'm new to this part of the code so you'll certainly have a better idea of what's going on here.
Could it happen because `joinWith` creates a tuple from what were top-level Rows?

cloud-fan replied:
Can you run the test locally, add some print here to see what's the `objDeserializer` that causes the bug?

cdegroc replied:
Oh sure! It's a CreateExternalRow Expression. If I `println(enc.objDeserializer)`, it prints
```
createexternalrow(getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._0.toString, getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._1, StructField(a,StringType,true), StructField(b,IntegerType,false))
createexternalrow(getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._0.toString, getcolumnbyordinal(0, StructField(a,StringType,true), StructField(b,IntegerType,false))._1, StructField(a,StringType,true), StructField(b,IntegerType,false))
```

cdegroc replied:
Unless there's a broader change to make, we could reduce the blast radius by:
- limiting the change to `CreateExternalRow` (i.e. check `enc.objDeserializer.isInstanceOf[CreateExternalRow]`)?
- having a dedicated tuple ExpressionEncoder for `Dataset.joinWith` (i.e. ~ update `ExpressionEncoder.tuple` to add a `nullSafe: Boolean = false` flag, set it to true for `Dataset.joinWith` and manually propagate nulls if true) ?

cloud-fan replied:
> having a dedicated tuple ExpressionEncoder for Dataset.joinWith

I like this idea. Ideally, for an `ExpressionEncoder[T]`, the deserialize expression assumes the input row is not null, as the input rows are from a `DataFrame` and Spark doesn't allow null as top-level rows. We can see from `ScalaReflection.deserializerForType` that we use `AssertNotNull` to guarantee it. This null check seems missing in `RowEncoder.deserializerFor`

In `ExpressionEncoder.tuple`, we inherit this assumption and assume each element of the Tuple can't be null. This assumption is broken in `joinWith`, because the Tuple element can be null for outer joins.

cloud-fan replied:
This bug only occurs for `RowEncoder`, as `Dataset[T].joinWith(Dataset[U])` works fine: https://github.com/apache/spark/pull/13425/files#diff-b98c99535d2b28cb47774860d500030e732c244c55b1ac05aead5d1cf1e7a602R772

Can you look into it and figure out the difference? This may help us the understand the bug better and guide us to the proper fix.

cdegroc replied:
The difference comes from the `RowEncoder` deserializer:
- When using a `Dataset[T]`, the `ExpressionEncoder` is used and calls [`ScalaReflection.deserializerForType`](https://github.com/apache/spark/blob/edc52857ca3ac031481a910f7871ff8d5e030e18/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala#L56) to get a deserializer for class `T`, which automatically [wraps the expression in a null-safe expression](https://github.com/apache/spark/blob/58d3f1516ed812b692709991e551829aa0090578/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L390-L394).
- When using a `DataFrame`, the `RowEncoder` is used and returns a [`CreateExternalRow` (not wrapped in a null-safe expression)](https://github.com/apache/spark/blob/58d3f1516ed812b692709991e551829aa0090578/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala#L254-L259).

I'm not sure there's an easy way to solve this, as the RowEncoder should guarantee (afaik) that top-level Rows aren't `null`.

Actually I think everything is already summarized in your initial PR that patched the tuple encoder to wrap deserializers in a null-safe way: https://github.com/apache/spark/pull/13425

cloud-fan replied:
> as the RowEncoder should guarantee (afaik) that top-level Rows aren't null.

This is not true for outer join. Shall we also add a null check to wrap `CreateExternalRow`?

cloud-fan replied:
OK let's update `joinWith` then.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:
```diff
@@ -110,23 +110,28 @@ object ExpressionEncoder {
+      val newDeserializer = enc.objDeserializer.transformUp {
         case GetColumnByOrdinal(0, _) => input
       }
+      if (enc.objSerializer.nullable) {
```
shouldn't we check `input.nullable`?

cdegroc replied:
`input` is a `GetStructField(GetColumnByOrdinal)` and calling nullable on it [will throw](https://github.com/apache/spark/blob/0a6be8cf59d5f9978337ec39953f0f7a88aec636/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala#L578) `org.apache.spark.sql.catalyst.analysis.UnresolvedException: Invalid call to nullable on unresolved object`.

I've used `enc.objSerializer.nullable` because that's what was used before [the regression](https://github.com/apache/spark/commit/cd92f25be5a221e0d4618925f7bc9dfd3bb8cb59).

cdegroc committed:
[SPARK-37829][SQL] Add if(isnull ...) check for DataFrame.joinWith

Wrap tuple fields deserializers in null checks when calling on
DataFrames as top-level rows are not nullable and won't propagate
null values.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala:
```diff
@@ -77,6 +77,28 @@ object RowEncoder {
+   */
+  private[sql] def nullSafe(exprEnc: ExpressionEncoder[Row]): ExpressionEncoder[Row] = {
+    val newDeserializerInput = GetColumnByOrdinal(0, exprEnc.objSerializer.dataType)
+    val newDeserializer: Expression = if (exprEnc.objSerializer.nullable) {
```
Sorry the code here is a bit confusing. We check `exprEnc.objSerializer.nullable` and then we construct `IsNull(newDeserializerInput)`? What's their connection?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1173,8 +1173,20 @@ class Dataset[T] private[sql](
+    // As we might be running on DataFrames, we need a custom encoder that will properly
+    // handle null top-level Rows.
+    def nullSafe[V](exprEnc: ExpressionEncoder[V]): ExpressionEncoder[V] = {
+      if (exprEnc.clsTag.runtimeClass != classOf[Row]) {
```
This looks a bit ugly.

> I've tried simply wrapping CreateExternalRow with a null check and a number of tests started failing as they were assuming top-level rows couldn't be null.

Are they UT or end-to-end tests? If they are UT, we can simply update the tests because we have changed the assumption.
